compiler/rustc_middle/src/ty/consts/kind.rs

```diff
@@ -68,6 +72,63 @@ pub enum ConstKind<'tcx> {
     /// A placeholder for a const which could not be computed; this is
     /// propagated to avoid useless error messages.
     Error(ty::DelaySpanBugEmitted),
+
+    /// Abstract which contains an expression which has partially evaluated items.
+    Abstract(Expr<'tcx>),
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Hash)]
+#[derive(HashStable, TyEncodable, TyDecodable)]
+pub enum Expr<'tcx> {
+    Binop(mir::BinOp, Const<'tcx>, Const<'tcx>),
+    UnOp(mir::UnOp, Const<'tcx>),
+    FunctionCall(Const<'tcx>, &'tcx [Const<'tcx>]),
+    Cast(CastKind, Const<'tcx>, Ty<'tcx>),
+}
```

Comment:
Uh-oh. I am extremely concerned about where this will lead, in particular seeing as there are function calls in that list here. It is in general *not true* that `f(expr)` is the same as first evaluating `expr` to a constant and then passing the result to `f`, since the round-trip through a valtree will reset padding. @lcnr filed an issue about this somewhere. *At least* this needs to be referenced here, but in general I am worried that we have (or seem to have) no overarching design, and no story for how any of this makes sense, and just "figure it out as we go" -- I think that's a recipe for disaster when it comes to type system design.

Reply:
> in particular seeing as there are function calls in that list here. It is in general not true that f(expr) is the same as first evaluating expr to a constant and then passing the result to f, since the round-trip through a valtree will reset padding

IMO generic consts should sort of "desugar" to a `const` item + substs i.e. `N + 1` -> `ADD::<N, 1_usize>`, and generic args I think cant preserve padding since valtrees are "lossy" that way. That makes this fine I think :thinking: (we dont do this kind of desugaring rn but long term I think its what we want, lcnr might disagree though :laughing:) 'desugaring' is probably not the right word. Im not sure if it's feasible to _actually_ desugar it, but i think its the semantics we'd want to emulate if we cant desguar.

> but in general I am worried that we have (or seem to have) no overarching design, and no story for how any of this makes sense, and just "figure it out as we go"

theres currently a WIP hackmd doc that's writing down a "long term" plan for generic_const_exprs and there's a pile of issues on the `project-const-generics` repo for "open design issues" and such, so I think eventually once this doc is written you might be satisfied ^^ (the padding stuff is something that would be mentioned in the doc)

I think this PR isnt really 'blocked' on any of this, its not changing any logic or adding new logic just emulating existing stuff (even if the existing stuff is a bit questionable and we should be figuring everything out more)

Reply:
> That makes this fine I think

No, it does not. I was specifically talking about function calls: `f(expr)`, where `expr` has a type like `&(i32, i8)`. Here, `f(expr)` and `f({const C = expr; C})` are *not equivalent*.

> its not changing any logic or adding new logic just emulating existing stuff (even if the existing stuff is a bit questionable and we should be figuring everything out more)

Where do we currently assume that the evaluation of `f(expr)` can be split up into first evaluating `expr` to a constant `C`, and then evaluating `f(C)`?

Reply:
> theres currently a WIP hackmd doc that's writing down a "long term" plan for generic_const_exprs and there's a pile of issues on the project-const-generics repo for "open design issues" and such, so I think eventually once this doc is written you might be satisfied ^^ (the padding stuff is something that would be mentioned in the doc)

I might be, but I am also extremely worried that a dependent type system is not something you can first implement and the document the design of. You need to know very well which design you want to go for before you start implementing things.

What you are implementing here is, to my knowledge, the first ever dependent type system for an unsafe language. I would say that's about a PhD thesis worth of theory material to be written down.

Reply:
Ah here is the existing issue for this: https://github.com/rust-lang/project-const-generics/issues/20.

Reply:
Okay, so looking at the diff, this seems to mirror the existing `abstract_const::Node` type? I won't ask why we now have the same tree structure twice... but I guess that is what you meant when you said that this is just refactoring existing code. Fair enough, I didn't realize that we are *already doing this dangerous thing*. :/ IMO that's just unsound. What's the syntax to construct such an abstract const?

Reply:
We should probably continue this in https://github.com/rust-lang/project-const-generics/issues/20.

Reply:
Currently, this refactors the existing way to create `AbstractConst`s through a builder defined in `compiler/rustc_ty_utils/src/consts.rs`. This looks through the thir and converts exprs recursively.

The reason this refactor is being performed is that the prior implementation lazily substituted parameters, and for generic associated consts such as
```
// Taken from zulip https://rust-lang.zulipchat.com/#narrow/stream/260443-project-const-generics/topic/directly.20embedding.20abstract.20consts.20in.20.60ty.3A.3AConst.60/near/290945289
trait Encode {
    const SIZE: usize;
    fn encode(&self) -> [u8; Self::SIZE];
}

impl<const N: usize> Encode for [u8; N] {
    const SIZE: usize = N;
    fn encode(&self) -> [u8; Self::SIZE] {
        *self
    }
}
```
we want to switch to a more eager substitution, which this does.